packages/block-editor/src/components/block-context/index.js

```diff
@@ -0,0 +1,37 @@
+/**
+ * WordPress dependencies
+ */
+import { createContext, useContext } from '@wordpress/element';
+
+/** @typedef {import('react').ReactNode} ReactNode */
+
+/**
+ * @typedef BlockContextProviderProps
+ *
+ * @property {Record<string,*>} value    Context value to merge with current
+ *                                       value.
+ * @property {ReactNode}        children Component children.
+ */
+
+/** @type {import('react').Context<Record<string,*>>} */
+const Context = createContext( {} );
+
+/**
+ * Component which merges passed value with current consumed block context.
+ *
+ * @see https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-context/README.md
+ *
+ * @param {BlockContextProviderProps} props
+ */
+export function BlockContextProvider( { value, children } ) {
+	const context = useContext( Context );
+
+	return (
+		<Context.Provider
+			value={ { ...context, ...value } }
```

Comment:
I think the issue with `value` in the provider still applies even when React hooks are used:

> Because context uses reference identity to determine when to re-render, there are some gotchas that could trigger unintentional renders in consumers when a provider’s parent re-renders. For example, the code below will re-render all consumers every time the Provider re-renders because a new object is always created for `value`

https://reactjs.org/docs/context.html#caveats

Reply:
> I think the issue with `value` in the provider still applies even when React hooks are used:

Related: https://github.com/WordPress/gutenberg/pull/21467#discussion_r405262284

We can probably memoize this a bit.

Reply:
> I think the issue with `value` in the provider still applies even when React hooks are used:

Implemented memoization / caching between d97e93612 and 9b96b6584 .